### PR TITLE
Implement functionality in purchase-order screen ( #1wmz45q ) 

### PR DESCRIPTION
**Pull request by meet-aniket**

### CODE REVIEW

File: src/store/modules/purchaseOrder/actions.ts [hunk 1]
```diff
@@ -11,19 +11,23 @@ import emitter from '@/event-bus'
...
     // if (payload.viewIndex === 0) emitter.emit("presentLoader");
 
     let resp;
 
     try {
-      resp = await PurchaseOrdersService.fetchPurchaseOrders()
+      resp = await PurchaseOrdersService.fetchPurchaseOrders(payload)
 
       // resp.data.response.numFound tells the number of items in the response
+      const orders = resp.data.grouped.orderId.groups
```
Comment (ymaheshwari1) [inline src/store/modules/purchaseOrder/actions.ts:25]: If the resp does not contain grouped then this line will through an error of `cannot read property grouped of undefined`. So, we will assign the resp to a variable after we have checked that the status is 200 and we are getting some data.

File: src/store/modules/purchaseOrder/actions.ts [hunk 2]
```diff
@@ -11,19 +11,23 @@ import emitter from '@/event-bus'
...
       if (resp.status === 200 && !hasError(resp)) {
```
Comment (ymaheshwari1) [inline src/store/modules/purchaseOrder/actions.ts:26]: If the resp.status is 200 but the data is empty then the state will not be updated as expected there might be an error. So add a condition to check whether we are getting some data with status 200 or not

File: src/views/PurchaseOrders.vue [hunk 3]
```diff
@@ -33,8 +32,33 @@ export default defineComponent({
...
+      orders: 'purchaseOrder/getPurchaseOrders',
+      totalOrders: 'purchaseOrder/getTotalPurchaseOrder'
+    })
+  },
+  methods: {
+    async getPurchaseOrder(){
+      const paylaod = {
+        "json": {
+          "params": {
+            "rows": 10,
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrders.vue:46]: Why we have hardcoded rows value, suppose if we have 50 PO then we will always get the first 10, need to make it dynamic.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/store/modules/purchaseOrder/actions.ts (hunk 1 above)
Comment (meet-aniket) [reply src/store/modules/purchaseOrder/actions.ts]: Improved code quality Sir.

File: src/store/modules/purchaseOrder/actions.ts (hunk 2 above)
Comment (meet-aniket) [reply src/store/modules/purchaseOrder/actions.ts]: Improved code Sir.

File: src/views/PurchaseOrders.vue (hunk 3 above)
Comment (meet-aniket) [reply src/views/PurchaseOrders.vue]: Made it dynamic Sir.

File: src/services/PurchaseOrderService.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+const fetchPurchaseOrders = async (payload: any): Promise <any>  => {
+  return api({
+   // TODO: We can replace this with any API
+    url: "/solr-query", 
+    method: "POST",
+    data: payload
+  });
+}
+
+const fetchPODetails = async (query: any): Promise <any> => {
```
Comment (ymaheshwari1) [inline src/services/PurchaseOrderService.ts:12]: Unwanted method declaration for this PR, we can remove this.

File: src/store/modules/purchaseOrder/actions.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    // Show loader only when new query and not the infinite scroll
+    // if (payload.viewIndex === 0) emitter.emit("presentLoader");
+
+    let resp;
+
+    try {
+      resp = await PurchaseOrdersService.fetchPurchaseOrders(payload)
+
+      // resp.data.response.numFound tells the number of items in the response
```
Comment (ymaheshwari1) [inline src/store/modules/purchaseOrder/actions.ts:24]: Remove unwanted comments

File: src/store/modules/purchaseOrder/actions.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      if (resp.status === 200 && !hasError(resp) && resp.data.grouped) {
+        const orders = resp.data.grouped.orderId.groups
+        commit(types.PURCHASE_ORDER_UPDATED, {
+          list: orders,
+          total: orders.length
```
Comment (ymaheshwari1) [inline src/store/modules/purchaseOrder/actions.ts:29]: The total value here should be the no of purchase orders, but here we are storing the length of orders received.

File: src/views/PurchaseOrders.vue [hunk 7]
```diff
@@ -0,0 +1,77 @@
+<template>
+  <ion-page>
+    <ion-header :translucent="true">
+      <ion-toolbar>
+        <ion-title>{{ $t("Purchase Orders") }}</ion-title>
+      </ion-toolbar>
+    </ion-header>
+    <ion-content :fullscreen="true">
+      <div>
+        <ion-searchbar :placeholder="$t('Search')" />
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrders.vue:10]: Implement search functionality, for now, search using the `orderId` and `externalOrderId` and confirm once with @dt2patel Sir for the searchable fields.

File: src/views/PurchaseOrders.vue [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  },
+  computed: {
+    ...mapGetters({
+      orders: 'purchaseOrder/getPurchaseOrders',
+      totalOrders: 'purchaseOrder/getTotalPurchaseOrder'
+    })
+  },
+  methods: {
+    async getPurchaseOrder(vRow?: any){
+      vRow = vRow ? vRow : process.env.VUE_APP_VIEW_ROW;
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrders.vue:43]:
```suggestion
      const vRow = vRow ? vRow : process.env.VUE_APP_VIEW_SIZE;
```

File: src/views/PurchaseOrders.vue [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      const paylaod = {
+        "json": {
+          "params": {
+            "rows": vRow,
+            "group": true,
+            "group.field": "orderId",
+            "group.limit": 10000
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrders.vue:50]: Add a param `"group.ngroups": true` in query which will return the total number of PO's

File: src/views/PurchaseOrders.vue [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        }
+      }
+      this.store.dispatch('purchaseOrder/findPurchaseOrders', paylaod);
+    }
+  },
+  mounted () {
+    this.getPurchaseOrder();
+  },
+  setup () {
+    const router = useRouter();
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrders.vue:65]: Remove router as it's not used in this component.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]
Implement infinite scrolling

File: src/views/PurchaseOrders.vue (hunk 7 above)
Comment (dt2patel) [reply src/views/PurchaseOrders.vue:10]: Ideally if we are getting PO from SOLR anything indexed know SOLR should be searchable right?

Comment (ymaheshwari1) [reply src/views/PurchaseOrders.vue:10]: Yes Sir

File: src/services/PurchaseOrderService.ts (hunk 4 above)
Comment (meet-aniket) [reply src/services/PurchaseOrderService.ts:12]: Removed Sir.

File: src/store/modules/purchaseOrder/actions.ts (hunk 5 above)
Comment (meet-aniket) [reply src/store/modules/purchaseOrder/actions.ts:24]: Removed Sir.

File: src/store/modules/purchaseOrder/actions.ts (hunk 6 above)
Comment (meet-aniket) [reply src/store/modules/purchaseOrder/actions.ts:29]: Changed this value from length to ngroups which comes in API response.

File: src/views/PurchaseOrders.vue (hunk 9 above)
Comment (meet-aniket) [reply src/views/PurchaseOrders.vue:50]: Added Sir.

File: src/views/PurchaseOrders.vue (hunk 8 above)
Comment (meet-aniket) [reply src/views/PurchaseOrders.vue:43]: Changed as per your suggestions.

File: src/views/PurchaseOrders.vue (hunk 10 above)
Comment (meet-aniket) [reply src/views/PurchaseOrders.vue]: Removed Sir.

File: src/views/PurchaseOrders.vue [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    }
+  },
+  computed: {
+    ...mapGetters({
+      orders: 'order/getPurchaseOrders',
+      isScrolleable: 'order/isScrolleable'
+    })
+  },
+  methods: {
+    async getPurchaseOrder(vSize?: any, vIndex?: any){
```
Comment (adityasharma7) [inline src/views/PurchaseOrders.vue:52]: Use consistent name for purchaseOrders

File: src/views/PurchaseOrders.vue [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      const viewSize = vSize ? vSize : process.env.VUE_APP_VIEW_SIZE;
+      const viewIndex = vIndex ? vIndex : 0;
+
+      const paylaod = {
```
Comment (adityasharma7) [inline src/views/PurchaseOrders.vue:56]:
```suggestion
      const payload = {
```

File: src/views/PurchaseOrders.vue [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+            "start": viewSize * viewIndex,
+            "group": true,
+            "group.field": "orderId",
+            "group.limit": 10000,
+            "group.ngroups": true,
+            "defType": "edismax",
+            "q.op": "AND",
+            "qf": "orderId"
+          },
+          "query": `(*${this.queryString}*) OR "${this.queryString}"^100`,
```
Comment (adityasharma7) [inline src/views/PurchaseOrders.vue:69]: You could exclude edimax if the querystring is not provided

Reference code
```
const payload = {     
        "json": {
          "params": {
            "rows": viewSize,
            "start": viewIndex
          } as any,
          "query": "*:*",
          "filter" : 'docType: PRODUCT AND isVirtual: false'   
        } 
      }
      if (this.queryString) {
        payload.json.query = '*' + this.queryString + '*';
        payload.json.params.defType = 'edismax';
        payload.json.params.qf = "upc productId sku groupId productName productCategoryNames";
      }
      ```

File: src/store/modules/order/mutation-types.ts [hunk 14]
```diff
@@ -0,0 +1,2 @@
+export const SN_ORDER = 'order'
+export const PURCHASE_ORDER_UPDATED = SN_ORDER + '/PURCHASE_ORDER_UPDATED'
```
Comment (adityasharma7) [inline src/store/modules/order/mutation-types.ts:2]: Please use consistent name

File: src/store/modules/order/actions.ts [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        const orders = resp.data.grouped.orderId
+        
+        if (payload.json.params.start && payload.json.params.start > 0) orders.groups = state.purchaseOrders.list.concat(orders.groups);
+        commit(types.PURCHASE_ORDER_UPDATED, {
+          list: orders.groups,
+          total: orders.ngroups
+        })
+      } else {
+        //showing error whenever not getting Orders
+        showToast(translate("Orders not found"));
```
Comment (adityasharma7) [inline src/store/modules/order/actions.ts:28]: I think we have not added it to en.json. Please ensure it for other text as well

File: src/components/PurchaseOrderItem.vue [hunk 16]
```diff
@@ -0,0 +1,43 @@
+<template>
+  <ion-item button @click="viewPurchaseOrder()" lines="none">
+    <ion-label>
+      <!-- TODO:- Handle this purchase order number property for now i have used OrderName or OrderId -->
+      <h3>{{ purchaseOrder.orderName ? purchaseOrder.orderName : purchaseOrder.orderId }}</h3>
+      <p>{{purchaseOrder.externalOrderId}}</p>
```
Comment (adityasharma7) [inline src/components/PurchaseOrderItem.vue:6]:
```suggestion
      <p>{{ purchaseOrder.externalOrderId }}</p>
```

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

File: src/store/modules/order/getters.ts [hunk 17]
```diff
@@ -0,0 +1,13 @@
+import { GetterTree } from "vuex";
+import OrderState from "./OrderState";
+import RootState from "../../RootState";
+
+const getters: GetterTree<OrderState, RootState> = {
+  getPurchaseOrders(state) {
+    return state.purchaseOrders.list;
+  },
+  isScrolleable(state) {
```
Comment (ymaheshwari1) [inline src/store/modules/order/getters.ts:9]:
```suggestion
  isScrollable(state) {
```

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/views/PurchaseOrders.vue [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+  <ion-page>
+    <ion-header :translucent="true">
+      <ion-toolbar>
+        <ion-title>{{ $t("Purchase Orders") }}</ion-title>
+      </ion-toolbar>
+    </ion-header>
+    <ion-content :fullscreen="true">
+      <div>
+        <ion-searchbar :placeholder="$t('Search')" v-model="queryString" @keyup.enter="getPurchaseOrders()" />
+      </div>
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrders.vue:11]: I think we don't need this div.

File: src/views/PurchaseOrders.vue [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+          "params": {
+            "rows": viewSize,
+            "start": viewSize * viewIndex,
+            "group": true,
+            "group.field": "orderId",
+            "group.limit": 10000,
+            "group.ngroups": true,
+          } as any,
+          "query": "*:*",
+          "filter": `docType: ORDER AND orderTypeId: PURCHASE_ORDER`
```
Comment (ymaheshwari1) [inline src/views/PurchaseOrders.vue:67]:
```suggestion
          "filter": 'docType: ORDER AND orderTypeId: PURCHASE_ORDER'
```

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/views/PurchaseOrders.vue [hunk 20]
```diff
@@ -0,0 +1,99 @@
+<template>
+  <ion-page>
+    <ion-header :translucent="true">
+      <ion-toolbar>
+        <ion-title>{{ $t("Purchase Orders") }}</ion-title>
+      </ion-toolbar>
+    </ion-header>
+    <ion-content :fullscreen="true">
```
Comment (dt2patel) [inline src/views/PurchaseOrders.vue:8]: Remove full screen

**Review by dt2patel — DISMISSED** [review]

File: src/store/modules/order/mutations.ts [hunk 21]
```diff
@@ -0,0 +1,11 @@
+import { MutationTree } from 'vuex'
+import OrderState from './OrderState'
+import * as types from './mutation-types'
+
+const mutations: MutationTree <OrderState> = {
+  [types.ORDER_PURCHASE_UPDATED] (state, payload) {
```
Comment (adityasharma7) [inline src/store/modules/order/mutations.ts:6]:
```suggestion
  [types.ORDER_PRCHS_ORDRS_UPDATED] (state, payload) {
```

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

**Review by ymaheshwari1 — APPROVED** [review]

**Review by adityasharma7 — APPROVED** [review]

**Pull request merged**